### PR TITLE
do type fixups for calls so that linkname can be used across types

### PR DESCRIPTION
**Pull request by jaddr2line**
This took longer than expected to implement, prerequisite to #606

### CODE REVIEW

**Comment by deadprogram** [comment]
As amusing as `darkarts.go` is for a file name, I think perhaps a more mundane and literal naming might be better. For example: `typefixes.go`.

**Comment by aykevl** [comment]
Or more general: `compiler-pragmas.go`.

**Comment by deadprogram** [comment]
@aykevl can you please review? Seems legit to me.

File: testdata/compiler-pragmas.go [hunk 1]
```diff
@@ -0,0 +1,36 @@
+package main
+
+import (
+	"unsafe"
+	"time"
+)
+
+//go:linkname getCoroutine runtime.getCoroutine
+func getCoroutine() unsafe.Pointer
```
Comment (aykevl) [inline testdata/compiler-pragmas.go]: I would prefer if this test only tested for the feature in question (`//go:linkname` with slightly mismatched types), without also relying on internals of the runtime. You can test for just this feature like this, for example (untested):

```go
var global = 3

//go:linkname foo foo
func foo(ptr unsafe.Pointer) {
    if ptr == unsafe.Pointer(&global) {
        println("pointer matches")
    } else {
        println("pointer mismatch!")
    }
}

//go:linkname foobar foo
func foobar(ptr *int)

func main() {
    foobar(&global)
}
```

File: compiler/calls.go [hunk 2]
```diff
@@ -38,6 +39,48 @@ func (c *Compiler) createCall(fn llvm.Value, args []llvm.Value, name string) llv
 		fragments := c.expandFormalParam(arg)
 		expanded = append(expanded, fragments...)
 	}
+	fnParams := fn.Type().ElementType().ParamTypes()
+pcheck:
+	if len(fnParams) != len(expanded) {
+		if len(fnParams) < len(expanded) {
+			// too many parameters
+			// attempt to trim off parent handle/etc. if not used
```
Comment (aykevl) [inline compiler/calls.go]: Is this actually needed right now? If not, I'd recommend just erroring here.

File: compiler/compiler.go [hunk 3]
```diff
@@ -1371,7 +1386,21 @@ func (c *Compiler) parseCall(frame *Frame, instr *ssa.CallCommon) (llvm.Value, e
...
+		switch retTSSA.Len() {
+		case 0:
+			retT = c.ctx.VoidType()
+		case 1:
+			retT = c.getLLVMType(retTSSA.At(0).Type())
+		default:
+			retT = c.getLLVMType(retTSSA)
+		}
+		if retT != ret.Type() {
+			ret = c.builder.CreateBitCast(ret, retT, "call.cast")
```
Comment (aykevl) [inline compiler/compiler.go]: A bitcast cannot be used for all values, such as structs (used for multiple returns).

Also, please add a test that mismatched return types work with more than one return value.

File: compiler/calls.go (hunk 2 above)
Comment (aykevl) [reply compiler/calls.go]: Can you give an example?

If this is for functions marked `//export`, then it is incorrect: they use a different ABI and may diverge in the future (but are nearly the same right now except for the last two parameters). `//export` tries to strictly follow the C ABI while TinyGo internally uses a different ABI that may be more efficient (and is not stable).

Comment (jadr2ddude) [reply compiler/calls.go]: @aykevl As far as I can tell, this only seems to happen when invoking runtime.yield via linkname on the tasks scheduler.

File: compiler/compiler.go [hunk 4]
```diff
@@ -337,8 +337,20 @@ func (c *Compiler) Compile(mainPath string) []error {
...
+			continue
+		}
+		frames = append(frames, c.parseFuncDecl(f))
+	}
+
+	// Declare all functions that do not have a body.
+	for _, f := range c.ir.Functions {
+		if f.Blocks != nil {
+			continue
+		}
```
Comment (aykevl) [inline compiler/compiler.go]: This does not appear to be necessary: the tests still pass without this change.

File: compiler/calls.go (hunk 2 above)
Comment (aykevl) [reply compiler/calls.go]: Also, the tests pass without this parameter stripping. I would suggest using the correct number of parameters across mismatching calls.

**Comment by aykevl** [comment]
LLVM should eventually make pointer types opaque, meaning that they are just of type "pointer" with no destination type. That would make this PR mostly unnecessary. However, it's work that has been going on for four years now so I wouldn't exactly expect that to happen soon.
More information: https://llvm.org/devmtg/2015-10/slides/Blaikie-OpaquePointerTypes.pdf

File: compiler/compiler.go (hunk 4 above)
Comment (jadr2ddude) [reply compiler/compiler.go]: I have confirmed that this is necessary. At the moment I don't really see a reasonable way to test this.

Basically, the interrupt synchronization primitive package on `intsched` demonstrates this behavior, as the `machine/sync` package gets initialized before `runtime`, which causes the compiler to crash later due to assertions. This is necessary to ensure that the functions are initialized in the correct order.

File: compiler/calls.go (hunk 2 above)
Comment (jadr2ddude) [reply compiler/calls.go]: Alright. I switched this to a cleaner (somewhat) mechanism.

**Comment by jadr2ddude** [comment]
> While this PR is relatively small, this change introduces quite a bit of complexity in the compiler that I would really like to avoid (the compiler is already more complex than I'd like).
Same.

>There are many places where the compiler assumes a certain signature, such as in func values.
This is a limitation here for at least the moment.

>Also, the order in which functions are declared (with vs without body) cannot in general be relied upon, especially once packages start being compiled independently.
Once packages start being compiled independently, this may be less of an issue. With the exception of the problem of linknaming exported functions, this should actually work by just linking the functions.

> * use `unsafe.Pointer` in both cases:
>   ```go
>   func foobar(ptr unsafe.Pointer)
>   ```
This would essentially mean that `*runtime.task` would need to replaced with an `unsafe.Pointer` in a wide variety of places. I would really prefer to keep types - in my opinion ditching all of the type information inside the `runtime` is worse.

> * use unnamed types in both cases:
>   ```go
>   func foobar(ptr *struct{a int, b string, c *byte})
>   ```
This isn't really feasible.

Additionally, we may need this behavior to make random go code work.

**Comment by jadr2ddude** [comment]
I am pushing this back to v0.11 because I don't want this going into v0.10 without us using it much. I will try to get a rebase in soon, and we will need to get this in towards the beginning of the v0.11 development cycle because it will be necessary for interrupt support.

**Comment by jaddr2line** [comment]
I have decided to go in a different direction.

**Pull request closed without merging**